### PR TITLE
Add moderation replies using menu

### PR DESCRIPTION
**Pull request by llewelld**
Allows a reply to be included with each moderation which is visible in the history table on the moderation page, and to the experience author on the edit story page.

In this implementation, moderators can see all previous replies in the table, but the author can see only the most recent reply.

This is a rather minimal approach to implementing what's described in #487, but could be expanded on later. It doesn't cover the rules and example responses portion of the task, those will still need to be added in.

### CODE REVIEW

**Comment by llewelld** [comment]
Converting to draft while I fix the failing tests.

**Comment by llewelld** [comment]
The tests now pass, so I'm moving this out of draft again. The coverage isn't great, but that's due to new lines in `share_experience()` and it looks like it wouldn't make sense trying to address these until pull requet #472 (or its successor) has been merged in.

File: static/css/main.css [hunk 1]
```diff
@@ -528,3 +528,34 @@ body {
...
 }
 
+.textbox-fixed {
+  overflow: auto;
+  height: 125px;
+  background-color: #e9ecef;
+  opacity: 1;
+}
+
+/* COLLAPSABLE CELLS ON THE MODERATION PAGE */
```
Comment (gedankenstuecke) [inline static/css/main.css:538]: This is a really nice touch! 👍

**Review by gedankenstuecke — APPROVED** [review]
This is great! Technically everything works as expected and I agree, that for the testing parts we can wait for the other PR with tests to be merged in at this point. 

There's a few front-end things we might want to improve, but not necessarily in this PR (also can use tomorrow's meet-up to gather feedback on this!).  

1. In the moderator view the two different reply types are not very obvious, maybe those could be made a bit more clear. 
2. On the `view my experience` page the reply does not show up (the one at `/main/view/<uuid>/`, while it shows up in the `edit` view)

Lastly, there's a potentially bigger piece of work, which would be to inform the user that their experience has been moderated when the status is set to approved/rejected. This could be done via email through the `oh_member.message()` function, which takes a subject & body as parameters and uses the OH API to request sending an email on our behalf through OH. But this could also be its own PR from my view!

**Comment by gedankenstuecke** [comment]
I've just gone ahead and solved a merge conflict in our css! :)

**Review by gedankenstuecke — COMMENTED** [review]
I left some comments/thoughts in the issue related to this pr (#487)

**Comment by llewelld** [comment]
Rebased.

**Comment by llewelld** [comment]
Rebased on top of #509.

PR #509 is now the "freetext" version of the changes. This version layers the moderation menu on top of it. I've updated the titles to reflect this.

**Comment by gedankenstuecke** [comment]
Thanks so much also in this thread, @llewelld! Just converted this and its sibling PR to drafts as they are open for discussion :)

**Comment by llewelld** [comment]
I've made some small changes (to make the menu options link to the code-of-conduct) in an additional commit in preparation for the monthly meeting tomorrow.

**Comment by llewelld** [comment]
Rebased.

**Comment by llewelld** [comment]
Rebased.

**Comment by llewelld** [comment]
For info (and ease of referece) this requires migration. To view the migration list:
```
docker exec -it autspaces-web-1 python manage.py showmigrations
```
To apply the migration for this PR:
```
docker exec -it autspaces-web-1 python manage.py migrate main 0010_experiencehistory_change_reply
```
To remove the migration when changing back to main branch:
```
docker exec -it autspaces-web-1 python manage.py migrate main 0009_auto_20230327_1226
```

File: server/apps/main/templates/main/moderate_experience.html
Comment (helendduncan) [inline server/apps/main/templates/main/moderate_experience.html:1]: Minor to the point of being pedantic - but is there a situation where the title might also need to be moderatable_text?

**Review by gedankenstuecke — APPROVED** [review]
There's only some very minor comments from my side on this, as it works great! 

I've left a comment on where we could implement data validation to address #524. 

Otherwise I noticed that we're missing a high-level `<div class="container">` in the `moderate_experiences.html` that should go around the `<section id="share-form">` to avoid the whole viewport running over its boundaries :-)

File: server/apps/main/helpers.py [hunk 2]
```diff
@@ -133,7 +133,7 @@ def get_oh_metadata(ohmember, uuid):
     file = [f for f in files if f["metadata"]["uuid"] == uuid]
 
     if len(file) == 0:
-        file = None
+        file = [None]
```
[the remark this answers is not in the document]
Comment (llewelld) [reply server/apps/main/helpers.py:136]: This is a personal thing as I find it easier to manage resource lifetimes if functions have only a single entry and exit point (`with` alleviates this, but isn't always available). If you prefer the early return here I'd be happy to change it though?

File: server/apps/main/helpers.py [hunk 3]
```diff
@@ -524,12 +531,43 @@ def filter_in_review(files):
...
+            structured = ""
+    return structured
+
+def get_latest_change_reply(experience_id):
+    change_reply = ""
+    changed_at = None
+    # We catch exceptions in case the experience doesn't exist
+    # in the local database
+    try:
+        model = PublicExperience.objects.get(experience_id=experience_id)
```
[the remark this answers is not in the document]
Comment (llewelld) [reply server/apps/main/helpers.py]: Thanks, this is a good point, and I'm glad you flagged it. I've updated the variable name.

File: server/apps/main/views.py [hunk 4]
```diff
@@ -198,10 +199,13 @@ def share_experience(request, uuid=False):
                 form = ShareExperienceForm(data)
                 title = "Edit experience"
                 moderation_status = data.get("moderation_status", "not reviewed")
+                change_reply, changed_at = get_latest_change_reply(uuid)
```
[the remark this answers is not in the document]
Comment (llewelld) [reply server/apps/main/views.py:207]: Nicely identified, thank you! I've added the replies to the invalid form path too.

File: server/apps/main/views.py [hunk 5]
```diff
@@ -457,6 +465,7 @@ def moderate_experience(request, uuid):
             data = {**unchanged_experience_details, **trigger_details}
 
             moderation_comments = data.pop("moderation_comments", None)
+            moderation_reply = data.pop("moderation_reply", "")
```
[the remark this answers is not in the document]
Comment (llewelld) [reply server/apps/main/views.py:473]: I've not made any changes here, hoping that this can be addressed by @mastoffel on #524.

**Comment by llewelld** [comment]
> Otherwise I noticed that we're missing a high-level `<div class="container">` in the `moderate_experiences.html` that should go around the `<section id="share-form">` to avoid the whole viewport running over its boundaries :-)

Nicely spotted! I've wrapped the contents of the section with a container div.

Thanks everyone for the really helpful reviews. I've updated the code and will now squash the commits and migrations. I'll message back here once that's completed.

**Comment by llewelld** [comment]
Commits squashed and rebased.

File: server/apps/main/helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gedankenstuecke) [reply server/apps/main/helpers.py:136]: No strong preference for either way, I was just genuinely curious in order to learn something myself, and that makes total sense 👍 🙂

File: server/apps/main/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gedankenstuecke) [reply server/apps/main/views.py:207]: great, thanks!

**Review by gedankenstuecke — APPROVED** [review]

**Comment by gedankenstuecke** [comment]
@llewelld Please feel free to push merge on this! 📌

**Pull request merged**